src/ipc/ipc4/helper.c

```diff
@@ -135,6 +135,16 @@ struct comp_dev *comp_new_ipc4(struct ipc4_module_init_instance *module_init)
 	return dev;
 }
 
+static struct ipc_comp_dev *get_comp(struct ipc *ipc, uint16_t type, uint32_t id)
+{
+	struct ipc_comp_dev *c = ipc_get_comp_by_id(ipc, id);
+
+	if (c && c->type == type)
```

Comment:
@andyross assuming there are 2 different types of components with the same ID, wouldnt this always fail if the first component in the list is not for the specified type even if  the other one exists in the list?

Reply:
It would, but by my reading that would be a bug in the design?  The list as currently used is already populated with polymorphic components, and already assuming that the first entry with a matching ID is the one desired.  Are there circumstances where it would be possible to have duplicate IDs?

Reply:
Ah...  OK, so if the component ID is input to the protocol, this actually is something subject to injected bugs that we'll need to address, I think.  I'll take a look, could be as simple as traversing the list and erroring out at creation time.

Reply:
Indeed, I see it now.  You're absolutely right, the initial component ID is set by the kernel, not the firmware.  So there needs to be protection in the firmware against duplicates, I'll see what I can work up and attach it to this PR, as it's related.  The fix here is protecting us against the results of the duplication (and even if there aren't duplicates, this remains an externally triggerable crash which we have to fix), but there is a rooter-cause that needs to be addressed too.